MLIBZ-2030: Fix Web Storage Adapters

thomasconner opened the pull request:
#### Description
This PR fixes several bugs that existed in the web storage adapters.

#### Changes
- Replace `lodash/findByIndex` with `[].indexOf()`.
- Fix `removeById()`.
- Retrieve correct substring for cookie storage.
- Fix typos referencing wrong storage adapters.

#### Tests
- Add unit tests

Commit:
Fix webstorage

Use indexOf instead of findIndex from lodash. Merge existing entities in storage for a collection when saving entities for the same collection. Use global.sessionStorage instead of global.localStorage for SessionStorageAdapter.

Commit:
Add chai and chai-as-promised as dev dependencies

Commit:
Fix removeById() for web storage adapters

Commit:
Add unit tests for webstorage adapters

gprodanov commented on src/middleware/src/storage/webstorage.js:
```diff
@@ -4,7 +4,6 @@ import keyBy from 'lodash/keyBy';
 import merge from 'lodash/merge';
```
I think the three storage adapters should be in separate files, in a separate folder. I think it should be a general rule - "one class, one file".

thomasconner replied:
I don't agree. I think if things are closely enough related then it is perfectly fine to have multiple classes in one file.

gprodanov replied:
But this makes navigation between files so much more difficult. If I want to check something in `LocalStorageAdapter` class, I need to use "find in files". Or know that it is in an arbitrarily named file. Right now, "webstorage.js" would be the last thing that comes to mind, when looking for a storage adapter. I'd try "localstorage", "storage", "adapter" and will not find it. In this case "storage" is in the name of the actual file, but not always.

And this is the case with many more files/classes.

thomasconner replied:
I like the way it is and would not like to see it changed. I like that all web storage adapters are grouped into the same file since they are related.

gprodanov replied:
They'd still be grouped, but in a folder. Has the same benefits, without the drawbacks.

Anyways, your call.

gprodanov commented on src/middleware/src/storage/webstorage.js:
```diff
@@ -103,8 +104,8 @@ export class LocalStorageAdapter extends WebStorageAdapter {
         delete entitiesById[id];
-        return this.save(collection, values(entitiesById))
-          .then(() => ({ count: 1 }));
+        global.localStorage.setItem(`${this.name}${collection}`, JSON.stringify(values(entitiesById)));
```
A level of abstraction seems necessary here - building the key in a separate function would make it more flexible and less error prone. A helper "set()" or maybe just "serialize()" function would also be useful (in the base class, to be able to avoid thinking about serialization (stringification)). This is code that can go in the common class.

gprodanov commented on test/unit/storage/cookiestorage.test.js:
```diff
@@ -0,0 +1,183 @@
+    });
+  });
+
+  // describe('clear()', () => {
```
Is this intentionally commented out?

thomasconner replied:
Yes

thomasconner replied:
The cookie storage mock was wrong. I fixed the mock and have uncommented this test.

gprodanov commented on test/unit/storage/localstorage.test.js:
```diff
@@ -0,0 +1,193 @@
+
+describe('LocalStorageAdapter', () => {
+  beforeEach(() => {
+    global.localStorage = storageMock();
```
Is it safe to modify native, global objects like this? And deleting it in the afterEach()... Why not use a spy?

Which brings me to a more general question: are these unit tests, or integration tests. Because they seem like integration tests to me, and what I wrote for the live service were all unit tests.

thomasconner replied:
I now put back the original value that was replaced by the mock after the tests are completed.

gprodanov commented on src/middleware/src/storage/webstorage.js:
```diff
@@ -182,7 +183,7 @@ export class SessionStorageAdapter extends WebStorageAdapter {
   save(collection, entities) {
```
Much of the code is identical across all three storage types. Basically the only largely different part is the `_find()`. The rest is almost identical.

thomasconner replied:
There are some more differences that would make this a bit more difficult. For now things work the way they are so lets leave it this way.

gprodanov replied:
[Here's a diff](http://www.mergely.com/TR03BEfQ/). Literally "localStorage" vs "sessionStorage" and the undefined check.

The cookie storage is almost the same, except it has the one more line copied a few times - about the expiration. We should log some technical debt tickets in JIRA.

gprodanov commented on test/unit/storage/cookiestorage.test.js:
```diff
@@ -0,0 +1,183 @@
+chai.use(chaiAsPromised);
+chai.should();
+
+function randomString() {
```
This function seems to be present in multiple locations, should move it to a utils file.

thomasconner replied:
I moved the function to `utils.js` and use it to shared `randomString()` across the unit tests.

gprodanov reviewed (commented):
My comments are not specific to this PR, they are general architecture questions, as is often the case. But we never really discuss them, they just hang there for a while and we move on.

So, as always, I am not requesting changes, I am starting a discussion. You guys can merge and continue, since it isn't the focus of this PR, but I do this because if we discuss it separately, I might struggle to think of examples, so I point them out in PRs.

thomasconner commented:
@georgiwe You would need to approve the PR to allow me to merge the PR without using my administrator privileges. Can you please do this if everything is fine?

gprodanov commented:
I submitted a review with comments earlier. I tried to explain my view in the general review comment. It's over [here](https://github.com/Kinvey/html5-sdk/pull/25#pullrequestreview-62999453).

Commit:
Move common randomString() to utils file so it can be shared.

Commit:
Fix unit tests

Use shared randomString() in utils. Replace mock storage with previous value when tests complete. Fix cookie storage mock.
